src/position.cpp

```diff
@@ -602,11 +611,12 @@ bool Position::pseudo_legal(const Move m) const {
   Piece pc = moved_piece(m);
 
   // Use a slower but simpler function for uncommon cases
-  if (type_of(m) != NORMAL || is_gating(m))
+  if (type_of(m) != NORMAL)
       return MoveList<LEGAL>(*this).contains(m);
-
-  // Is not a promotion, so promotion piece must be empty
-  if (promotion_type(m) - KNIGHT != NO_PIECE_TYPE)
+  
+  // If the move gates a piece make sure we have that piece in hand
+  // and that we are allowed to gate on the from square.
+  if (gating_type(m) != KING && !(in_hand(us, gating_type(m)) && (gates(us) & from)))
```

Comment:
If I understand correctly, changing it to `if (gating_type(m) != KING && !(gating_type(m) < PIECE_TYPE_NB && in_hand(us, gating_type(m)) && (gates(us) & from)))` would allow us to get rid of the +1 in the size of the `inHand` array, since this is the only place where `gating_type` can be `PIECE_TYPE_NB`, right?

Reply:
Yes that sounds right to me. A more ambitious idea would be to define HAWK and ELEPHANT after ROOK but before QUEEN. In that case these kind of hacks would disappear entirely and it would make it a lot easier to implement the Seirawan2880 idea from the other discussion. I imagine it would be tricky to get this working as a non-functional change though. On the plus side if that did happen I have in mind some ways to make the move encoding nicer. See here for the outline of my idea:

https://gist.github.com/ceebo/a03403823605e11b0b08568aed732526

It would involve getting rid of all statements of the form type_of(move) == TYPE and using is_type(move) instead. Anyway, it's all speculation at the moment... !

Reply:
Yes, my plan also was to move HAWK and ELEPHANT after ROOK or QUEEN at some later point, but a lot of changes will be necessary, so for now I try to avoid it.

Since we only deal with one variant here, we have more freedom regarding the move encoding, so in principle I like your idea, but I would still like to keep the move types, since they are used all over the place in the code and if we want to keep merging upstream changes, that would cause a lot of merge conflicts.